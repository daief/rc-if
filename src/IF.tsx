import * as React from 'react';

export interface CommonProps {
  show?: boolean;
}

export interface RcIF extends React.SFC<CommonProps> {
  If: React.SFC<{ if?: boolean }>;
  ElseIf: React.SFC<{ elseIf?: boolean }>;
  Else: React.SFC;
}

export const __NAME_IF__ = 'IF_name_if';
export const __NAME_ELSE_IF__ = 'IF_name_else_if';
export const __NAME_ELSE__ = 'IF_name_else';

export const ctx = React.createContext<boolean>(false);

export const Wrap: React.SFC<CommonProps> = ({ show = true, children }) => {
  if (!show) {
    return null;
  }

  return (
    <ctx.Consumer>
      {isInnerCtx => {
        if (!isInnerCtx) {
          return null;
        }

        const array = React.Children.toArray(children);
        let hasPassed = false;
        let hasIf = false;

        const renderArray = array.map(child => {
          // @ts-ignore
          if (child && child.type && child.props) {
            // @ts-ignore
            const { props: childProps } = child;
            // @ts-ignore
            const { displayName } = child.type;
            switch (displayName) {
              case __NAME_IF__: {
                hasIf = true;
                hasPassed = !!childProps.if;
                return hasPassed ? child : null;
              }
              case __NAME_ELSE_IF__: {
                if (!hasIf) {
                  throw new Error('Wrong logical order.');
                }

                if (hasPassed) {
                  return null;
                }
                hasPassed = !!childProps.elseIf;
                return hasPassed ? child : null;
              }
              case __NAME_ELSE__: {
                if (!hasIf) {
                  throw new Error('Wrong logical order.');
                }
                // mark end
                hasIf = false;
                return hasPassed ? null : child;
              }
              default:
                return child;
            }
          }
          return child;
        });

        return <>{renderArray}</>;
      }}
    </ctx.Consumer>
  );
};

/**
 * FI is as a wrapper component. `FI` is used to distinguish it from `If`.
 */
export const FI: RcIF = props => (
  <ctx.Provider value={true}>
    <Wrap show={props.show}>{props.children}</Wrap>;
  </ctx.Provider>
);

FI.If = ({ children }) => <Wrap show>{children}</Wrap>;
FI.If.displayName = __NAME_IF__;
FI.ElseIf = ({ children }) => <Wrap show>{children}</Wrap>;
FI.ElseIf.displayName = __NAME_ELSE_IF__;
FI.Else = ({ children }) => <Wrap show>{children}</Wrap>;
FI.Else.displayName = __NAME_ELSE__;
