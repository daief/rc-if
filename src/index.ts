import { FI } from './IF';

const { If, ElseIf, Else } = FI;

export { FI, If, ElseIf, Else };
