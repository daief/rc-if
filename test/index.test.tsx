import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { FI, If, ElseIf, Else } from '../src';

const getComponent = (content = 'content'): React.SFC => () => <div>{content}</div>;

const renderComponent = (content = 'content') => React.createElement(getComponent(content));

describe('FI wrap', () => {
  function renderTest(show: any) {
    const C = getComponent();
    const component = renderer.create(
      <FI show={show}>
        <C />
      </FI>,
    );
    return {
      content: C,
      component,
    };
  }

  it('FI.show => unset', () => {
    const { component, content } = renderTest(undefined);
    expect(component.root.findByType(content).type).toBe(content);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('FI.show => true', () => {
    const { component, content } = renderTest(true);
    expect(component.root.findByType(content).type).toBe(content);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('FI.show => false', () => {
    const { component, content } = renderTest(false);
    expect(() => component.root.findByType(content)).toThrow();
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('FI & If', () => {
  it('If.if => unset', () => {
    const component = renderer.create(
      <FI>
        <If>{renderComponent()}</If>
      </FI>,
    );

    expect(() => component.root.findByType(If)).toThrow();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('If.if => true', () => {
    const component = renderer.create(
      <FI>
        <If if>{renderComponent()}</If>
      </FI>,
    );

    expect(component.root.findByType(If).type).toBe(If);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('FI & If & ElseIf & Else', () => {
  it('If.if => unset, ElseIf.elseIf => unset, Else', () => {
    const component = renderer.create(
      <FI>
        <If>{renderComponent('if')}</If>
        <ElseIf>{renderComponent('else if')}</ElseIf>
        <Else>{renderComponent('else')}</Else>
      </FI>,
    );

    expect(() => component.root.findByType(If)).toThrow();
    expect(() => component.root.findByType(ElseIf)).toThrow();
    expect(component.root.findByType(Else)).toBeTruthy();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('If.if => true, ElseIf.elseIf => true, Else', () => {
    const component = renderer.create(
      <FI>
        <If if>{renderComponent('if')}</If>
        <ElseIf elseIf>{renderComponent('else if')}</ElseIf>
        <Else>{renderComponent('else')}</Else>
      </FI>,
    );

    expect(() => component.root.findByType(ElseIf)).toThrow();
    expect(() => component.root.findByType(Else)).toThrow();
    expect(component.root.findByType(If)).toBeTruthy();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('If.if => false, ElseIf.elseIf => true, Else', () => {
    const component = renderer.create(
      <FI>
        <If>{renderComponent('if')}</If>
        <ElseIf elseIf>{renderComponent('else if')}</ElseIf>
        <Else>{renderComponent('else')}</Else>
      </FI>,
    );

    expect(() => component.root.findByType(If)).toThrow();
    expect(() => component.root.findByType(Else)).toThrow();
    expect(component.root.findByType(ElseIf)).toBeTruthy();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('If & ElseIf & Else, wrong order', () => {
    const spyFn = spyOn(console, 'error');

    const renderArray = [
      () =>
        renderer.create(
          <FI>
            <ElseIf elseIf />
            <If />
          </FI>,
        ),
      () =>
        renderer.create(
          <FI>
            <Else />
            <If />
          </FI>,
        ),
      () =>
        renderer.create(
          <FI>
            <ElseIf />
          </FI>,
        ),
      () =>
        renderer.create(
          <FI>
            <Else />
          </FI>,
        ),
      () =>
        renderer.create(
          <FI>
            <If />
            <ElseIf />
            <Else />
            <ElseIf />
          </FI>,
        ),
    ];

    renderArray.forEach(renderATest => {
      expect(renderATest).toThrowError(/Wrong logical order/);
    });

    expect(spyFn).toHaveBeenCalled();
    expect(spyFn).toHaveBeenCalledTimes(renderArray.length);
  });

  it('Without FI, render nothing', () => {
    const CArray = Array(4)
      .fill(1)
      .map(() => getComponent());
    const [C1, C2, C3, C4] = CArray;

    const component = renderer.create(
      <>
        <ElseIf>
          <C1 />
        </ElseIf>
        <Else>
          <C2 />
        </Else>
        <If if>
          <C3 />
        </If>
        <C4 />
      </>,
    );

    CArray.forEach((C, i) => {
      if (i < 3) {
        expect(() => component.root.findByType(C)).toThrow();
      } else {
        expect(component.root.findByType(C).type).toBe(C);
      }
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('Multiply', () => {
  it('Parallel', () => {
    const CArray = Array(4)
      .fill(1)
      .map((_, i) => getComponent(`content - ${i + 1}`));
    const [C1, C2, C3, C4] = CArray;
    const component = renderer.create(
      <FI>
        <If if>
          <C1 />
        </If>
        <ElseIf elseIf>{renderComponent('else if')}</ElseIf>

        <If>{renderComponent('if')}</If>
        <ElseIf>{renderComponent('else if')}</ElseIf>
        <ElseIf elseIf>
          <C2 />
        </ElseIf>
        <Else>{renderComponent('else')}</Else>

        <If if>
          <C3 />
        </If>

        <If>{renderComponent('if')}</If>
        <ElseIf>{renderComponent('else if')}</ElseIf>
        <ElseIf>{renderComponent('else if')}</ElseIf>
        <Else>
          <C4 />
        </Else>
      </FI>,
    );

    CArray.forEach(C => {
      expect(component.root.findByType(C).type).toBe(C);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Nested', () => {
    const CArray = Array(4)
      .fill(1)
      .map((_, i) => getComponent(`content - ${i + 1}`));
    const [C1, C2, C3, C4] = CArray;
    const component = renderer.create(
      <FI>
        <If />
        <ElseIf />
        <Else>
          <If if>
            <C1 />
          </If>
          <If />
          <ElseIf elseIf>
            <C2 />
            <If />
            <Else>
              <If />
              <Else>
                <C3 />
                <If if>
                  <If if>
                    <If if>
                      <If if>
                        <If />
                        <Else>
                          <C4 />
                        </Else>
                      </If>
                    </If>
                  </If>
                </If>
              </Else>
            </Else>
          </ElseIf>
        </Else>
      </FI>,
    );

    CArray.forEach(C => {
      expect(component.root.findByType(C).type).toBe(C);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});
