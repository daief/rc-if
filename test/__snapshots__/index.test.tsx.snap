// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FI & If & ElseIf & Else If.if => false, ElseIf.elseIf => true, Else 1`] = `
<div>
  else if
</div>
`;

exports[`FI & If & ElseIf & Else If.if => true, ElseIf.elseIf => true, Else 1`] = `
<div>
  if
</div>
`;

exports[`FI & If & ElseIf & Else If.if => unset, ElseIf.elseIf => unset, Else 1`] = `
<div>
  else
</div>
`;

exports[`FI & If & ElseIf & Else Only FI(or If, ElseIf, Else) wraps and is in parent and child will render 1`] = `
Array [
  <div>
    <div>
      content
    </div>
  </div>,
  <div>
    content
  </div>,
]
`;

exports[`FI & If & ElseIf & Else Without FI, render nothing 1`] = `
<div>
  content
</div>
`;

exports[`FI & If If.if => true 1`] = `
<div>
  content
</div>
`;

exports[`FI & If If.if => unset 1`] = `null`;

exports[`FI wrap FI.show => false 1`] = `null`;

exports[`FI wrap FI.show => true 1`] = `
<div>
  content
</div>
`;

exports[`FI wrap FI.show => unset 1`] = `
<div>
  content
</div>
`;

exports[`Multiply Nested 1`] = `
Array [
  <div>
    content - 1
  </div>,
  <div>
    content - 2
  </div>,
  <div>
    content - 3
  </div>,
  <div>
    content - 4
  </div>,
]
`;

exports[`Multiply Parallel 1`] = `
Array [
  <div>
    content - 1
  </div>,
  <div>
    content - 2
  </div>,
  <div>
    content - 3
  </div>,
  <div>
    content - 4
  </div>,
]
`;
